Simplify and fix videos module structure

1zaman opened the pull request:
Restructures the "My Videos" module to nest the `PlayerFragment` inside `MyRecentVideosFragment`, which fixes issues with the former's state retention. Changes manual `TabHost` implementation to `ViewPager` and `TabLayout`. Detailed explanations can be found in the commit messages.

Resolves [MA-1820][] and [MA-1050][].

[MA-1820]: https://openedx.atlassian.net/browse/MA-1820
[MA-1050]: https://openedx.atlassian.net/browse/MA-1050

bguertin commented:
@1zaman I get a crash in Recent Videos if I:
1) Play a video
2) Rotate to Landscape
3) Rotate to portrait
4) Try to play a different video
5) CRASH
```
FATAL EXCEPTION: main
                                                              Process: org.edx.mobile, PID: 6757
                                                              java.lang.NullPointerException: Attempt to invoke interface method 'java.lang.String org.edx.mobile.model.VideoModel.getVideoId()' on a null object reference
                                                                  at org.edx.mobile.view.MyRecentVideosFragment.playVideoModel(MyRecentVideosFragment.java:272)
                                                                  at org.edx.mobile.view.MyRecentVideosFragment.access$300(MyRecentVideosFragment.java:49)
                                                                  at org.edx.mobile.view.MyRecentVideosFragment$1.onItemClick(MyRecentVideosFragment.java:101)
                                                                  at org.edx.mobile.view.adapters.MyRecentVideoAdapter$1.onClick(MyRecentVideoAdapter.java:54)
                                                                  at android.view.View.performClick(View.java:4785)
                                                                  at android.view.View$PerformClick.run(View.java:19884)
                                                                  at android.os.Handler.handleCallback(Handler.java:739)
                                                                  at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                  at android.os.Looper.loop(Looper.java:135)
                                                                  at android.app.ActivityThread.main(ActivityThread.java:5343)
                                                                  at java.lang.reflect.Method.invoke(Native Method)
                                                                  at java.lang.reflect.Method.invoke(Method.java:372)
                                                                  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:905)
                                                                  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:700)
```

bguertin commented:
This fix doesn't address the same issues in `VideoListFragment`

bguertin commented:
Actually, this doesn't seem to fix the issue at all, even in the Recent Videos fragment; when I rotate the screen, there are no captions or next/prev buttons

bguertin commented:
I think it would be nice if we used a ViewPager here, to enable swipe gestures and easy integration with `TabLayout` which we should be using for a more modern (Material) design.

miankhalid commented:
encountered the same crash as Brian mentioned. Other than that issue mentioned in MA-1963 seems to be fixed :star2:

1zaman commented:
I can't reproduce this crash following those steps, though I don't have my device connected to the network at the moment. In any case, I have implemented a simpler patch for just the inconsistent state (and not the state retention) at #546, which might be used instead for this release as it has less potential for new bugs.

1zaman commented:
Actually I managed to reproduce it now. Fixing.

1zaman commented:
Fixed in 26efd60028d1c50eea2d5ff3f211f6f38afc5edb. I am also implementing a `ViewPager` and `TabLayout` structure, which is mostly done except for freezing the `ViewPager` in landscape and theming the `TabLayout`.

@bguertin, @BenjiLee: Not sure of the exact scenario of [the PlayerFragment crash report](https://fabric.io/edx-inc/android/apps/org.edx.mobile/issues/56a9f7c7f5d3a7f76b66600d), since the the current `PlayerFragment` behaviour is to have it's instance retained, but it's probably also fixed by 26efd60028d1c50eea2d5ff3f211f6f38afc5edb. If you like, you can merge that commit into the release branch.

1zaman commented:
@bguertin: I have implemented the `ViewPager` and `TabLayout` design, resolving [MA-1050][]. Feel free to take another look. cc @miankhalid, @BenjiLee

[MA-1050]: https://openedx.atlassian.net/browse/MA-1050

bguertin commented:
@1zaman Heh, you've been waiting months to fix this :wink: 

Reviewing now

bguertin commented:
Looks pretty good, I'm having a problem though, steps:
1) In `Recent Videos`, select a video to play
2) While the video is playing, switch to `All Videos` tab
3) Tap on a course in the `All Videos` tab
4) Press back button to return to `My Videos`
5) Switch to `Recent Videos` tab
6) Notice that the video player appears to be broken; the video I had started is broken and I can't tap on other videos to play them, it just shows a loading spinner indefinitely.

![screenshot_2016-02-03-10-27-40](https://cloud.githubusercontent.com/assets/1685613/12786930/2e1c69f2-ca61-11e5-93bb-2ca82648f8cc.png)

miankhalid commented:
Great work with the tabs and view pager implementation :star2: 

An issue that i found, is as follows:
1. Tap on RECENT VIDEOS.
2. Play any video.
3. Let the video end (hint: seek to almost end).
4. When the video ends the Play button shows up.
5. Change the orientation and play the video, sounds plays and video doesn't.

Same issue noticed in ALL VIDEOS tab as well.

BenjiLee commented:
That tab switching animation is so nice :3

BenjiLee commented:
Status bar is showing when I'm in fullscreen. 

Using an emulator on marshmallow

BenjiLee commented:
When playing fullscreen in "All Videos", pressing the back button will exit fullscreen. When playing fullscreen "Recent Videos", pressing the back button send you to My Courses regardless of the the last place you were.

BenjiLee commented:
What is MyAllVideos supposed to mean?

1zaman commented:
> Heh, you've been waiting months to fix this :wink:

I logged [MA-1050][] as a result of a discussion with Akiva in #303, but we never got around to working on this module before now.

[MA-1050]: https://openedx.atlassian.net/browse/MA-1050

1zaman commented:
@bguertin: I have fixed the issue with the video being automatically resumed on coming back to `MyVideosTabActivity` in 69bea3cc183999ed134a923a228d58ff78a1a786. The media player still gets stuck when I use the "Don't keep activities" option, although I can't reproduce it in debug mode on my device (I guess it might be a race condition of some kind). I'll look into it later.

1zaman commented:
@miankhalid: I can't reproduce the issue you reported.

1zaman commented:
@BenjiLee: Yeah, the back button finishes the Activity. This behaviour is the same on all screens where the media player is implemented. To go back to portrait mode you can rotate the screen, or toggle the full-screen button in the bottom left corner.

1zaman commented:
I've created [MA-2059][] for the media player state restoration brokenness. Since this pull request seems to be otherwise fine, and this issue doesn't seem to be directly related to or caused by it, I'm merging this now.

[MA-2059]: https://openedx.atlassian.net/browse/MA-2059

BenjiLee commented:
obligatory :+1:
